File: apps/ios/FishKit/Sources/PersonalChat/Views/message-voice-player/VoicePlaybackSpeed.swift
import Foundation

enum VoicePlaybackSpeed: Float, CaseIterable, Identifiable, Sendable {
    case slow = 0.75
    case normal = 1.0
    case fast = 1.5
    case veryFast = 2.0

    static let userDefaultsKey = "fish.voicePlaybackSpeed"

    var id: Float { rawValue }

    var label: String {
        switch self {
        case .slow: "0.75×"
        case .normal: "1×"
        case .fast: "1.5×"
        case .veryFast: "2×"
        }
    }

    var accessibilityLabel: String {
        switch self {
        case .slow: "0.75 times speed"
        case .normal: "Normal speed"
        case .fast: "1.5 times speed"
        case .veryFast: "2 times speed"
        }
    }

    static var persisted: Self {
        let stored = UserDefaults.standard.float(forKey: userDefaultsKey)
        guard stored > 0 else { return .normal }
        return allCases.min { abs($0.rawValue - stored) < abs($1.rawValue - stored) } ?? .normal
    }

    func persist() {
        UserDefaults.standard.set(rawValue, forKey: Self.userDefaultsKey)
    }
}

File: apps/android/feature/chat/src/main/kotlin/space/fishhub/android/feature/chat/VoicePlaybackSpeed.kt
package space.fishhub.android.feature.chat

import android.content.Context

internal enum class VoicePlaybackSpeed(val multiplier: Float, val label: String) {
    Slow(0.75f, "0.75×"),
    Normal(1.0f, "1×"),
    Fast(1.5f, "1.5×"),
    VeryFast(2.0f, "2×");

    val accessibilityLabel: String
        get() = when (this) {
            Slow -> "0.75 times speed"
            Normal -> "Normal speed"
            Fast -> "1.5 times speed"
            VeryFast -> "2 times speed"
        }

    companion object {
        private const val PreferencesName = "fish-voice-preferences"
        private const val SpeedKey = "playback_speed"

        fun persisted(context: Context): VoicePlaybackSpeed {
            val stored = context.applicationContext
                .getSharedPreferences(PreferencesName, Context.MODE_PRIVATE)
                .getFloat(SpeedKey, Normal.multiplier)
            return entries.minByOrNull { kotlin.math.abs(it.multiplier - stored) } ?: Normal
        }

        fun persist(context: Context, speed: VoicePlaybackSpeed) {
            context.applicationContext
                .getSharedPreferences(PreferencesName, Context.MODE_PRIVATE)
                .edit()
                .putFloat(SpeedKey, speed.multiplier)
                .apply()
        }
    }
}
